src/util/range_stream.js

```diff
@@ -0,0 +1,54 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+var stream = require('stream');
+
+/**
+ *
+ * RangeStream
+ *
+ * A transforming stream that returns range of the input.
+ *
+ */
+class RangeStream extends stream.Transform {
+
+    // The input end is exclusive
+    constructor(start, end, options) {
+        super(options);
+        this._start = start;
+        this._end = end;
+        this._size = end - start;
+        this._recvd_bytes = 0;
+        this._sent_bytes = 0;
+    }
+
+    /**
+     * implement the stream's Transform._transform() function.
+     */
+    _transform(data, encoding, callback) {
```

Comment:
I think that a stream implementation calling this.end() is delicate because this will close the writable end of the stream unexpectedly. A safer way is this.push(null) to close the readable end of the transform stream.
Also for a bit better clarity I would avoid mutating _start and _end and use a _pos offset.
So I would write it like so:
```
_transform(data, encoding, callback) {
    const r = range_utils.intersection(this._start - this._pos, this._end - this._pos, 0, data.length);
    if (r) this.push(data.slice(r.start, r.end));
    this._pos += data.length;
    if (this._pos >= this._end) this.push(null);
    return callback();
}
```

Reply:
I think your suggested implementation is good.